internal/datanode/channel_manager.go

```diff
@@ -0,0 +1,501 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type releaseFunc func(channel string)
+
+type ChannelManager struct {
+	dn *DataNode
+
+	communicateCh     chan *opState
+	runningFlowgraphs *flowgraphManager
+	opRunners         *typeutil.ConcurrentMap[string, *opRunner] // channel -> runner
+	abnormals         *typeutil.ConcurrentMap[int64, string]     // OpID -> Channel
+
+	releaseFunc releaseFunc
+
+	closeCh     chan struct{}
+	closeOnce   sync.Once
+	closeWaiter sync.WaitGroup
+}
+
+func NewChannelManager(dn *DataNode) *ChannelManager {
+	fm := newFlowgraphManager()
+	cm := ChannelManager{
+		dn: dn,
+
+		communicateCh:     make(chan *opState, 100),
+		runningFlowgraphs: fm,
+		opRunners:         typeutil.NewConcurrentMap[string, *opRunner](),
+		abnormals:         typeutil.NewConcurrentMap[int64, string](),
+
+		releaseFunc: fm.release,
+
+		closeCh: make(chan struct{}),
+	}
+
+	return &cm
+}
+
+func (m *ChannelManager) Submit(info *datapb.ChannelWatchInfo) error {
+	channel := info.GetVchan().GetChannelName()
+	runner := m.getOrCreateRunner(channel)
+	return runner.Enqueue(info)
+}
+
+func (c *ChannelManager) GetProgress(info *datapb.ChannelWatchInfo) *datapb.ChannelOperationProgressResponse {
```

Comment:
Do we need some locks here to ensure concurrency for reading and writing?

Reply:
cool, we could add a lock between finishOp and GetProgress